### ISSUE TITLE
ERR: fork/exec /dev/shm/rtspss-embeddedexe: no such file or directory

### ISSUE THREAD

**Issue by shadmanhabib** [aler9/rtsp-simple-server#1584]
## Which version are you using?

v0.21.6_linux_arm64v8 - downloaded from the release page.

## Which operating system are you using?

- [ ] Linux amd64 standard
- [ ] Linux amd64 Docker
- [x] Linux arm64 standard - Raspberry Pi OS Lite 64-bit Bullseye
- [ ] Linux arm64 Docker
- [ ] Linux arm7 standard
- [ ] Linux arm7 Docker
- [ ] Linux arm6 standard
- [ ] Linux arm6 Docker
- [ ] Windows amd64 standard
- [ ] Windows amd64 Docker (WSL backend)
- [ ] macOS amd64 standard
- [ ] macOS amd64 Docker
- [ ] Other (please describe)

## Describe the issue

Unable to run RTSP server. Log shows missing files/directories. 

Camera is recognised by libcamera and is able to save photos and videos.

## Describe how to replicate the issue

1. start the server
2. publish with rpiCamera - YAML file set as below
3. read with vlc - unable to be read

## Did you attach the server logs?

yes [rtsp-simple-server.log](https://github.com/aler9/rtsp-simple-server/files/11019254/rtsp-simple-server.log)

    2023/03/21 01:21:35 INF MediaMTX / rtsp-simple-server v0.21.6
    2023/03/21 01:21:35 DEB [path stream1] created
    2023/03/21 01:21:35 DEB path manager created
    2023/03/21 01:21:35 INF [path stream1] [rpicamera source] started
    2023/03/21 01:21:35 INF [RTSP] listener opened on :8554 (TCP), :8000 (UDP/RTP), :8001 (UDP/RTCP)
    2023/03/21 01:21:35 INF [path stream1] [rpicamera source] ERR: fork/exec /dev/shm/rtspss-embeddedexe-1679322095901196259: no such file or directory
    2023/03/21 01:21:40 INF [path stream1] [rpicamera source] ERR: fork/exec /dev/shm/rtspss-embeddedexe-1679322100911125333: no such file or directory
    2023/03/21 01:21:45 INF [path stream1] [rpicamera source] ERR: fork/exec /dev/shm/rtspss-embeddedexe-1679322105919164159: no such file or directory
    2023/03/21 01:21:46 INF shutting down gracefully
    2023/03/21 01:21:46 INF [path stream1] [rpicamera source] stopped
    2023/03/21 01:21:46 DEB [path stream1] destroyed (terminated)
    2023/03/21 01:21:46 INF [RTSP] listener is closing
    2023/03/21 01:21:46 DEB path manager is shutting down
    2023/03/21 01:21:46 INF waiting for external commands

## Did you attach a network dump?

no

###############################################
    # General parameters

    # Sets the verbosity of the program; available values are "error", "warn", "info", "debug".
    logLevel: debug
    # Destinations of log messages; available values are "stdout", "file" and "syslog".
    logDestinations: [stdout, file]
    # If "file" is in logDestinations, this is the file which will receive the logs.
    logFile: rtsp-simple-server.log

    # Timeout of read operations.
    readTimeout: 10s
    # Timeout of write operations.
    writeTimeout: 10s
    # Number of read buffers.
    # A higher value allows a wider throughput, a lower value allows to save RAM.
    readBufferCount: 512

    ###############################################
    # RTSP parameters

    # Disable support for the RTSP protocol.
    rtspDisable: no
    # List of enabled RTSP transport protocols.
    # UDP is the most performant, but doesn't work when there's a NAT/firewall between
    # server and clients, and doesn't support encryption.
    # UDP-multicast allows to save bandwidth when clients are all in the same LAN.
    # TCP is the most versatile, and does support encryption.
    # The handshake is always performed with TCP.
    protocols: [udp, multicast, tcp]
    # Encrypt handshakes and TCP streams with TLS (RTSPS).
    # Available values are "no", "strict", "optional".
    encryption: "no"
    # Address of the TCP/RTSP listener. This is needed only when encryption is "no" or "optional".
    rtspAddress: :8554
    # Address of the TCP/TLS/RTSPS listener. This is needed only when encryption is "strict" or "optional".
    # rtspsAddress: :8322
    # Address of the UDP/RTP listener. This is needed only when "udp" is in protocols.
    rtpAddress: :8000
    # Address of the UDP/RTCP listener. This is needed only when "udp" is in protocols.
    rtcpAddress: :8001
    # IP range of all UDP-multicast listeners. This is needed only when "multicast" is in protocols.
    multicastIPRange: 224.1.0.0/16
    # Port of all UDP-multicast/RTP listeners. This is needed only when "multicast" is in protocols.
    multicastRTPPort: 8002
    # Port of all UDP-multicast/RTCP listeners. This is needed only when "multicast" is in protocols.
    multicastRTCPPort: 8003
    # Path to the server key. This is needed only when encryption is "strict" or "optional".
    # This can be generated with:
    # openssl genrsa -out server.key 2048
    # openssl req -new -x509 -sha256 -key server.key -out server.crt -days 3650
    # serverKey: server.key
    # Path to the server certificate. This is needed only when encryption is "strict" or "optional".
    # serverCert: server.crt
    # Authentication methods.
    authMethods: [basic, digest]

    ###############################################
    # RTMP parameters

    # Disable support for the RTMP protocol.
    rtmpDisable: yes

    ###############################################
    # HLS parameters

    # Disable support for the HLS protocol.
    hlsDisable: yes

    ###############################################
    # WebRTC parameters

    # Disable support for the WebRTC protocol.
    webrtcDisable: yes

    ###############################################
    # Path parameters

    # These settings are path-dependent, and the map key is the name of the path.
    # It's possible to use regular expressions by using a tilde as prefix.
    # For example, "~^(test1|test2)$" will match both "test1" and "test2".
    # For example, "~^prefix" will match all paths that start with "prefix".
    # The settings under the path "all" are applied to all paths that do not match
    # another entry.
    paths:
      stream1:
        source: rpiCamera

        # ID of the camera
        rpiCameraCamID: 0
        # width of frames
        rpiCameraWidth: 1920
        # height of frames
        rpiCameraHeight: 1080
        # flip horizontally
        rpiCameraHFlip: false
        # flip vertically
        rpiCameraVFlip: false
        # brightness [-1, 1]
        # rpiCameraBrightness: 0
        # contrast [0, 16]
        # rpiCameraContrast: 1
        # saturation [0, 16]
        # rpiCameraSaturation: 1
        # sharpness [0, 16]
        # rpiCameraSharpness: 1
        # exposure mode.
        # values: normal, short, long, custom
        rpiCameraExposure: normal
        # auto-white-balance mode.
        # values: auto, incandescent, tungsten, fluorescent, indoor, daylight, cloudy, custom
        rpiCameraAWB: auto
        # denoise operating mode.
        # values: off, cdn_off, cdn_fast, cdn_hq
        rpiCameraDenoise: "off"
        # tuning file
        # rpiCameraTuningFile:
        # frames per second
        rpiCameraFPS: 25
        # period between IDR frames
        rpiCameraIDRPeriod: 60
        # bitrate
        rpiCameraBitrate: 1000000
        # H264 profile
        rpiCameraProfile: main
        # H264 level
        rpiCameraLevel: '4.1'

**Comment by aler9** [comment]
Hello, make sure that libcamera is installed correctly and up to date, by following instructions here:
https://github.com/aler9/rtsp-simple-server#from-a-raspberry-pi-camera

**Comment by shadmanhabib** [comment]
Libcamera seems to be all good, running v0.0.4. Legacy camera support has also been disabled through `raspi-config`.

    pi@alarm-pi:~ $ libcamera-still --version
    libcamera-apps build: 54a781dffdd1 01-03-2023 (17:47:29)
    libcamera build: v0.0.4+22-923f5d70

**Comment by shadmanhabib** [comment]
I get the same result as you when I run `mount | grep shm`.

pi@alarm-pi:~ $ mount | grep shm
tmpfs on /dev/shm type tmpfs (rw,nosuid,nodev)

This might be unrelated info but I also couldn't get `golang-go` working. I kept getting missing file errors even through I could see the executable in `/usr/local/go/bin`. From my quick research this seems like it happens when the build version and os version don't match.

To add insult to injury, I thought I had installed Raspberry Pi OS Lite 32-bit. Raspberry Pi Imager even says that version is the one cached on my computer. But the OS responds with `aarch64` with `name -m`. Any ideas?

**Comment by shadmanhabib** [comment]
You got it, all camera apps are showing up as `armhf`. I can uninstall all the camera apps but reinstalling seems to just go install the `armhf` version again. 

Running,

    apt list | grep camera

returns a number of libcamera apps but most of them show `armhf`, a couple of `all` but nothing specific to `arm64`. The packages labelled `all` seem to just be docs but not actual camera packages.

How can I force install an `arm64` version or am I stuck with having to recompile libcamera or reinstalling my OS?

**Comment by aler9** [comment]
In my opinion, you'd better reinstall the OS from scratch. As you wrote, you installed the 32bit version but ended up with a 64bit kernel and you have packages with mixed architectures.
Anyway, this is not a server issue. The server works with all the Raspberry Pi OS variants available for download.

**Comment by shadmanhabib** [comment]
Yeah that’s what I ended up doing. 

I ran `uname -a` between every few changes on the new install of Raspberry Pi OS Lite 32-bit and I kid you not I watched it change to 64-bit. Absolutely crazy. 

At this point I worked out it was either some `apt update` or one of the config settings from my Waveshare CM4 IO board.

I reinstalled the OS and didn’t update and used a different config for the IO board and it looks like we’re all sweet. Most likely culprit is Waveshare I think.

Thanks for all the advice @aler9!
